pkg/trace/agent/agent.go

```diff
@@ -202,105 +208,185 @@ func (a *Agent) loop() {
+	if filteredByTags(pt.Root, a.conf.RequireTags, a.conf.RejectTags) {
+		m.Source.TracesFiltered.Inc()
+		m.Source.SpansFiltered.Add(int64(len(pt.TraceChunk.Spans)))
+		return fmt.Errorf("Trace rejected as it fails to meet tag requirements. root: %v", pt.Root)
+	}
+	return nil
+}
 
-		// Root span is used to carry some trace-level metadata, such as sampling rate and priority.
-		root := traceutil.GetRoot(chunk.Spans)
-		normalizeChunk(chunk, root)
-		if !a.Blacklister.Allows(root) {
-			log.Debugf("Trace rejected by ignore resources rules. root: %v", root)
-			ts.TracesFiltered.Inc()
-			ts.SpansFiltered.Add(tracen)
-			p.RemoveChunk(i)
-			continue
+func eachSpan(fs ...SpanTransformer) Transformer {
+	return func(m *api.Metadata, pt *traceutil.ProcessedTrace) error {
+		for _, span := range pt.TraceChunk.Spans {
+			for _, f := range fs {
+				if err := f(m, pt, span); err != nil {
+					return err
+				}
+			}
 		}
+		return nil
+	}
+}
 
-		if filteredByTags(root, a.conf.RequireTags, a.conf.RejectTags) {
-			log.Debugf("Trace rejected as it fails to meet tag requirements. root: %v", root)
-			ts.TracesFiltered.Inc()
-			ts.SpansFiltered.Add(tracen)
-			p.RemoveChunk(i)
-			continue
+func (a *Agent) applyGlobalTags(m *api.Metadata, pt *traceutil.ProcessedTrace, s *pb.Span) error {
+	for k, v := range a.conf.GlobalTags {
+		if k == tagOrigin {
+			pt.TraceChunk.Origin = v
+		} else {
+			traceutil.SetMeta(s, k, v)
 		}
+	}
+	return nil
+}
 
-		// Extra sanitization steps of the trace.
-		for _, span := range chunk.Spans {
-			for k, v := range a.conf.GlobalTags {
-				if k == tagOrigin {
-					chunk.Origin = v
-				} else {
-					traceutil.SetMeta(span, k, v)
-				}
-			}
-			if a.ModifySpan != nil {
-				a.ModifySpan(span)
-			}
-			a.obfuscateSpan(span)
-			Truncate(span)
-			if p.ClientComputedTopLevel {
-				traceutil.UpdateTracerTopLevel(span)
-			}
-		}
-		a.Replacer.Replace(chunk.Spans)
+func (a *Agent) modifySpan(m *api.Metadata, pt *traceutil.ProcessedTrace, s *pb.Span) error {
+	if a.ModifySpan != nil {
+		a.ModifySpan(s)
+	}
+	return nil
+}
+
+func updateTopLevel(m *api.Metadata, pt *traceutil.ProcessedTrace, s *pb.Span) error {
+	if m.ClientComputedTopLevel {
+		traceutil.UpdateTracerTopLevel(s)
+	}
+	return nil
+}
 
-		a.setRootSpanTags(root)
-		if !p.ClientComputedTopLevel {
-			// Figure out the top-level spans now as it involves modifying the Metrics map
-			// which is not thread-safe while samplers and Concentrator might modify it too.
-			traceutil.ComputeTopLevel(chunk.Spans)
+func computeTopLevel(m *api.Metadata, pt *traceutil.ProcessedTrace) error {
+	if !m.ClientComputedTopLevel {
+		// Figure out the top-level spans now as it involves modifying the Metrics map
+		// which is not thread-safe while samplers and Concentrator might modify it too.
+		traceutil.ComputeTopLevel(pt.TraceChunk.Spans)
+	}
+	return nil
+}
+
+func (a *Agent) makeTransformers() []Transformer {
+	return []Transformer{
+		normalizePayloadEnv,
```

Comment:
This is cute, but personally I don't think it makes things easier to follow. I'd say it makes them _harder_ to follow. In this form, the attention is scattered between this list (which reflects the order in which things happen), and the meaning of its items. To understand the items and the order that they happen in, we now have to go to their definition and back to figure things out, resulting in multiple open tabs (or splits) to make sense of things. Furthermore, there seems to be additional API introduced (`eachSpan`) which requires learning even more concepts.

I may be biased, knowing the code for so long, but I think that the Process function was fine as it was. Yes, it was long, but it was clear to read, the order of things was obvious, and so was what was happening. The trace would come in at the top, and leave at the bottom.

Was this discussed in your team and you've concluded that it warrants a refactor? This is the first time I hear about this function being a problem. The way I see it is that, previously: one big function which clearly illustrates what happens in the trace-agent to a trace (and it used to be the "go to" point even for newcomers to get started), versus now, in this PR: a bunch of other functions, logic, API and concepts that scatter the actual processing. What was previously simple is now complex.

Reply:
We spoke about this, but reiterating for the PR record:

This was discussed with the team before and after the implementation and it's something we think makes the processing code clearer. One data point to promote this is that I found a redundant processing step (disposing of empty traces) repeated in the processing code, which was not obvious before the refactor.